hack/test-go.sh

```diff
@@ -1,129 +1,262 @@
-  if [[ "${1}" == *"${suffix}" ]]; then
-    # if the package name ends with /... they are intending a recursive test
-    test_packages=$(find_test_dirs "${1%$suffix}")
-  else
-    test_packages="$OS_GO_PACKAGE/$1"
-  fi
-elif [ -n "$TEST_KUBE" ]; then
-  test_packages=`find_test_dirs "."; special_upstream_test_dirs; find_upstream_test_dirs`
-else
-  test_packages=`find_test_dirs "."; special_upstream_test_dirs`
+# Build arguments for 'go test'
+if [[ -z "${verbose}" && -n "${junit_report}" ]]; then
+    # verbosity can be set explicitly by the user or set implicitly by asking for the jUnit 
+    # XML report, so we only want to add the flag if it hasn't been added by a user already 
+    # and is being implicitly set by jUnit report generation
+    gotest_flags+=" -v"
 fi
 
-if [ -n "$PRINT_PACKAGES" ]; then
-  for package in $test_packages
-  do
-    echo $package
-  done
+if [[ -n "${detect_races}" ]]; then
+    gotest_flags+=" -race"
+fi
 
-  exit 0
+# check to see if user has not disabled coverage mode
+if [[ -n "${coverage_spec}" ]]; then
+    # if we have a coverage spec set, we add it. '-race' implies '-cover -covermode atomic'
+    # but specifying both at the same time does not lead to an error so we can add both specs
+    gotest_flags+=" ${coverage_spec}"
 fi
 
-export OPENSHIFT_ON_PANIC=crash
+# check to see if user has not disabled test timeouts
+if [[ -n "${test_timeout}" ]]; then
+    gotest_flags+=" -timeout ${test_timeout}"
+fi
 
-if [[ -n "${KUBE_COVER}" && -n "${OUTPUT_COVERAGE}" ]]; then
-  # Iterate over packages to run coverage
-  test_packages=( $test_packages )
-  for test_package in "${test_packages[@]}"
-  do
-    mkdir -p "$OUTPUT_COVERAGE/$test_package"
-    KUBE_COVER_PROFILE="-coverprofile=$OUTPUT_COVERAGE/$test_package/profile.out"
+# list_test_packages_under lists all packages containing Golang test files that we want to run as unit tests
+# under the given base dir in the OpenShift Origin tree
+function list_test_packages_under() {
+    local basedir=$@
 
-    go test $KUBE_RACE -timeout $KUBE_TIMEOUT $KUBE_COVER "$KUBE_COVER_PROFILE" "$test_package" "${@:2}"
-  done
+    # we do not quote ${basedir} to allow for multiple arguments to be passed in as well as to allow for
+    # arguments that use expansion, e.g. paths containing brace expansion or wildcards
+    find ${basedir} -not \(                   \
+        \(                                    \
+              -path 'Godeps'                  \
+              -o -path '*_output'             \
+              -o -path '*_tools'              \
+              -o -path '*.git'                \
+              -o -path '*openshift.local.*'   \
+              -o -path '*Godeps/*'            \
+              -o -path '*assets/node_modules' \
+              -o -path '*test/*'              \
+        \) -prune                             \
+    \) -name '*_test.go' -print0 | xargs -0n1 dirname -z | sort -uz | xargs -0n1 printf "${OS_GO_PACKAGE}/%s\n"
```

Comment:
@smarterclayton this is the pipe I'm talking about. The `print0` long predates me. Before this pull it actually didn't propagate the behavior along the entire pipe so I fixed it to do that. You can see the older version above. Maybe we don't really want this printing shenanigans anyway since it was broken for so long and didn't impact anything.